clang/lib/CodeGen/Targets/AMDGPU.cpp

```diff
@@ -537,7 +537,11 @@ AMDGPUTargetCodeGenInfo::getLLVMSyncScopeID(const LangOptions &LangOpts,
     break;
   }
 
-  if (Ordering != llvm::AtomicOrdering::SequentiallyConsistent) {
+  // OpenCL assumes by default that atomic scopes are per-address space for
+  // non-sequentially consistent operations.
+  if (Scope >= SyncScope::OpenCLWorkGroup &&
```

Comment:
Can you explain which scopes will generate the one-as versions with this change? I don't know which ones are in that range: agent, workgroup, wavefront, singlethread are the possible candidates. 

I also don't understand why the one-as should be generated for a specific range of scopes. How do we know that will not lead to mis-compilation?

Reply:
OpenCL ones.

Reply:
Ok, so here's the enum. 

enum class SyncScope {
  SystemScope,
  DeviceScope,
  WorkgroupScope,
  WavefrontScope,
  SingleScope,
  HIPSingleThread,
  HIPWavefront,
  HIPWorkgroup,
  HIPAgent,
  HIPSystem,
  OpenCLWorkGroup,
  OpenCLDevice,
  OpenCLAllSVMDevices,
  OpenCLSubGroup,
  Last = OpenCLSubGroup
};

So like you said, the OpenCL ones. If Matt agrees, I am fine with this change though I don't see how it is better than explicitly checking for OpenCL. Just more cryptic. Either way, please add a comment.

Reply:
There's already a comment though?

Reply:
It's better because it's up to the OpenCL builtin, you can still use the 'standard' builtins in OpenCL which would then give them different behavior depending on the language.

Reply:
> There's already a comment though?

oh, ok, I see the comment now. Somehow missed it earlier.

Reply:
I think we should just stop emitting -one-as in any situation and rip out all the backend support after

Reply:
I'm fine with whatever makes C/C++ correct, OpenCL seems to do their own thing so I figured it's less friction to just leave it there.

Reply:
> less friction to just leave it there.

This is how we just end up with an ever-growing pile of garbage nobody understands. Just rip it out.

Reply:
> > less friction to just leave it there.
> 
> This is how we just end up with an ever-growing pile of garbage nobody understands. Just rip it out.

We'd better make sure we're not breaking OpenCL conformance before we do any ripping.

Reply:
Can we leave ripping it out completely until after we fix the issues that this causes for C/C++?

Reply:
The primary concern here would be performance loss